org.spoofax.terms/src/org/spoofax/terms/io/SimpleTextTermWriter.java

```diff
@@ -0,0 +1,410 @@
+     *
+     * @param term         the term whose annotations to write
+     * @param writer       the writer to write to
+     * @param depth        the current depth remaining
+     * @param isAnnotation whether the current term is an annotation term
+     * @throws IOException an I/O exception occurred
+     */
+    protected void writeAnnotations(IStrategoTerm term, Appendable writer, int depth, boolean isAnnotation) throws IOException {
+        IStrategoList annos = term.getAnnotations();
+        if (depth <= 1 || annos.isEmpty()) {
+            return;
+        }
+
+        writer.append('{');
+        writeSubterms(annos, writer, depth - 1, true);
+        writer.append('}');
+    }
+
+    /**
+     * Writes the attachments of a term.
+     *
+     * @param term         the term whose attachments to write
+     * @param writer       the writer to write to
+     * @param depth        the current depth remaining
+     * @param isAnnotation whether the current term is an annotation term
+     * @throws IOException an I/O exception occurred
+     */
+    protected void writeAttachments(IStrategoTerm term, Appendable writer, int depth, boolean isAnnotation) throws IOException {
+        ITermAttachment attachment = term.getAttachment(null);
+        if (depth <= 1 || attachment == null) {
+            return;
+        }
+
+        writer.append('«');
+        writeAttachment(attachment, writer, depth - 1);
+        attachment = attachment.getNext();
+        while (attachment != null) {
+            writer.append(LIST_SEPARATOR);
+            writeAttachment(attachment, writer, depth - 1);
+            attachment = attachment.getNext();
+        }
+        writer.append('»');
+    }
+
+    /**
+     * Writes a term attachment.
+     *
+     * @param attachment the term attachment to write
+     * @param writer     the writer to write to
+     * @param depth      the current depth remaining
+     * @throws IOException an I/O exception occurred
+     */
+    protected void writeAttachment(ITermAttachment attachment, Appendable writer, int depth) throws IOException {
+        writer.append(attachment.getClass().getSimpleName());
+        writer.append('<');
+        writer.append(attachment.toString());
+        writer.append('>');
+    }
+
+    /**
+     * Writes a list of subterms.
+     *
+     * @param term         the term whose subterms to write
+     * @param writer       the writer to write to
+     * @param depth        the current depth remaining for the subterms
+     * @param isAnnotation whether the current term is an annotation term
+     * @throws IOException an I/O exception occurred
+     */
+    private void writeSubterms(IStrategoTerm term, Appendable writer, int depth, boolean isAnnotation) throws IOException {
+        if (term.getSubtermCount() > 0) {
+            writeTerm(term.getSubterm(0), writer, depth, isAnnotation);
+            for (int i = 1; i < term.getSubtermCount(); i++) {
```

Comment:
I used `getSubtermCount()` and `getSubterm()` with indices because it avoids creating an iterator. In most cases this just delegates to `internalArray.length` and `internalArray[index]`. Is this not correct?

Reply:
In most cases, yes. But `StrategoList` doesn't have that.

Reply:
Changed to Iterator